src/components/genes/GenesSubscriber.js

```diff
@@ -1,75 +1,85 @@
-import React, { useState, useEffect, useCallback } from 'react';
+import React, {
+  useState, useEffect, useCallback, useMemo,
+} from 'react';
 import PubSub from 'pubsub-js';
-
-import Genes from './Genes';
-
-import TitleInfo from '../TitleInfo';
 import {
-  GENES_ADD,
-  CELLS_COLOR,
-  CLEAR_PLEASE_WAIT,
-  RESET,
+  EXPRESSION_MATRIX_ADD, CELLS_COLOR, CLEAR_PLEASE_WAIT, RESET,
 } from '../../events';
-import { interpolateColors } from '../utils';
+import { interpolatePlasma } from '../interpolate-colors';
+import { fromEntries, pluralize } from '../utils';
+import TitleInfo from '../TitleInfo';
+import Genes from './Genes';
 
 export default function GenesSubscriber(props) {
   const {
     onReady,
-    mapping,
     removeGridComponent,
-    variablesLabelOverride,
+    variablesLabelOverride: variablesLabel = 'gene',
     theme,
   } = props;
-  const [genes, setGenes] = useState({});
+
+  const [expressionMatrix, setExpressionMatrix] = useState();
   const [selectedId, setSelectedId] = useState(null);
   const [urls, setUrls] = useState([]);
 
   const onReadyCallback = useCallback(onReady, []);
 
   useEffect(() => {
-    const genesAddToken = PubSub.subscribe(
-      GENES_ADD, (msg, { data, url }) => {
-        setGenes(data);
-        setUrls((prevUrls) => {
-          const newUrls = [...prevUrls].concat({ url, name: 'Genes' });
-          return newUrls;
+    const expressionMatrixAddToken = PubSub.subscribe(
+      EXPRESSION_MATRIX_ADD, (msg, { data }) => {
+        const [attrs, arr] = data;
+
+        arr.get([null, null]).then((X) => {
```

Comment:
Yes I will try this. Although do you know off of the top of your head, if I run `.getRaw` multiple times, will the ArrayBuffer behind it change? 

Normally I would not expect / want it to change, but in this case I do, since I need to transfer the ArrayBuffer to the worker thread (in the Heatmap component), while here in the GenesSubscriber I always want to be able to access the data

Reply:
When you call `get` or `getRaw`, the first thing zarr does is create a new `RawArray` or `NestedArray` internally and then populates that array with chunk data.

https://github.com/gzuidhof/zarr.js/blob/f9252cd775d8fbe39bc0c5ab64bd37bdb2106a10/src/core/index.ts#L323

These constructors initialize a new empty buffer (or buffers for nested)

Reply:
A zarr array when initialized has _no_ data in memory, just metadata about the array. A zarr array is an interface to read in selections of a remote source lazily when calling `get` or `getRaw`. 

When you provide a selection to `get` or `getRaw`, zarr determines which chunks of source to read into memory and then assemble into an in-memory `NestedArray` or `RawArray`. So if you want to get all the data, I'd recommend calling `getRaw([null, null])` once and then making copies of the `RawArray.data` using `slice`. 

```

class Loader {
  constructor(url) {
    this.url = url;
    this.data = null;
  } 
  async get() {
    if (this.data) {
      return this.data.slice(); // copy
    }
    const z = await openArray({ store: this.url });
    const { data } = await z.getRaw([null, null]);
    this.data = data;
    return data;
  }
}
```

Reply:
Thanks, this is working!

Reply:
I have updated to use getRaw rather than get everywhere in https://github.com/hubmapconsortium/vitessce/pull/689/commits/81371a6da6d6188c6f677ba957655f12ed5f6154